p2p/transport/webrtc/listener.go

```diff
@@ -0,0 +1,350 @@
+package libp2pwebrtc
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/network"
+
+	tpt "github.com/libp2p/go-libp2p-core/transport"
+	"github.com/libp2p/go-libp2p/p2p/security/noise"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/multiformats/go-multibase"
+	"github.com/multiformats/go-multihash"
+
+	"github.com/pion/dtls/v2/pkg/crypto/fingerprint"
+	"github.com/pion/ice/v2"
+	"github.com/pion/webrtc/v3"
+)
+
+var (
+	ErrDataChannelTimeout    = errors.New("timed out waiting for datachannel")
+	ErrNoiseHandshakeTimeout = errors.New("noise handshake timeout")
+	ErrNoCertInConfig        = errors.New("no certificate configured in listener config")
+)
+
+var (
+	defaultMultihash *multihash.DecodedMultihash = nil
```

Comment:
What is this?

Reply:
Please see added comment

Reply:
I still don't understand why this is needed in the first place. This seems like an ugly hack. We should structure the code such that this isn't needed.